src/ripple/basics/spinlock.h

```diff
@@ -0,0 +1,160 @@
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright 2021, Nikolaos D. Bougalis <nikb@bougalis.net>
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+
+#ifndef RIPPLE_BASICS_PACKED_SPINLOCK_H_INCLUDED
+#define RIPPLE_BASICS_PACKED_SPINLOCK_H_INCLUDED
+
+#include <atomic>
+#include <limits>
+#include <thread>
+#include <type_traits>
+
+#ifndef __aarch64__
+#include <immintrin.h>
+#endif
+
+namespace ripple {
+
+namespace detail {
+/** Inform the processor that we are in a tight spin-wait loop.
+
+    Spinlocks caught in tight loops can result in the processor's pipeline
+    filling up with comparison operations, resulting in a misprediction at
+    the time the lock is finally acquired, necessitating pipeline flushing
+    which is ridiculously expensive and results in very high latency.
+
+    This function instructs the processor to "pause" for some architecture
+    specific amount of time, to prevent this.
+ */
+inline void
+spin_pause() noexcept
+{
+#ifdef __aarch64__
+    asm volatile("yield");
+#else
+    _mm_pause();
+#endif
+}
+
+}  // namespace detail
+
+/** A lock that grabs a single packed spinlock from an atomic integer
+
+    The implementation tries to use portable constructs but has to be
+    low-level for performance reasons.
+ */
+template <
+    class T,
+    class = std::enable_if_t<
+        std::is_unsigned_v<T> && std::atomic<T>::is_always_lock_free,
+        std::void_t<
+            decltype(std::declval<std::atomic<T>&>().fetch_and(0)),
+            decltype(std::declval<std::atomic<T>&>().fetch_or(0))>>>
+class packed_spinlock
+{
+private:
+    std::atomic<T>& bits_;
+    T mask_;
+
+public:
+    packed_spinlock(packed_spinlock const&) = delete;
+    packed_spinlock&
+    operator=(packed_spinlock const&) = delete;
+
+    packed_spinlock(std::atomic<T>& lock, int index)
+        : bits_(lock), mask_(1 << index)
+    {
+        assert(index >= 0 && (mask_ != 0));
+    }
+
+    [[nodiscard]] bool
+    try_lock()
+    {
+        return (bits_.fetch_or(mask_, std::memory_order_acquire) & mask_) == 0;
+    }
+
+    void
+    lock()
+    {
+        while (!try_lock())
+            detail::spin_pause();
```

Comment:
The primary justification was to simplify the codepath. There's also some weird (very low-level) interactions of `spin_pause` and comparisons that might actually serve to increase latency, which I was hoping to avoid.

I suspect that the `yield` isn't going to really help, but at the very least I ought to try and collect some additional metrics.